[c10d] Introduce a util for detecting DMA connectivity among devices

yifuwang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #129510

This PR introduces `_detect_dma_connectivity` - a utility for detecting DMA connectivity among devices.

The "DMA connectivity" in this context is more stringent than the ability to perform memory copy without CPU involvement. We define it as the ability for a device to issue load/store instructions and perform atomic operations on memory that resides on connected devices. The ability translates to the ability to run most aten GPU operations with operands backed by remote memory. `_detect_dma_connectivity` can help PyTorch and its users to determine whether certain DMA-based optimizations are possible.

`_detect_dma_connectivity` takes a `(device_type, connection_type)` pair and returns a matrix describing the connectivity. Connectivity detectors are statically registered on a `(device_type, connection_type)` basis. This PR implements the detector for `(CUDA, "nvlink")`. Later, detectors for pairs such as `(ROCM, "infinity_fabric")` can be introduced.

Example:

```python3
>>> from torch._C._autograd import DeviceType
>>> from torch._C._distributed_c10d import _detect_dma_connectivity
>>> connectivity = _detect_dma_connectivity(DeviceType.CUDA, "nvlink")
>>> for row in connectivity.matrix:
...     print(row)
...
[0, 18, 18, 18, 18, 18, 18, 18]
[18, 0, 18, 18, 18, 18, 18, 18]
[18, 18, 0, 18, 18, 18, 18, 18]
[18, 18, 18, 0, 18, 18, 18, 18]
[18, 18, 18, 18, 0, 18, 18, 18]
[18, 18, 18, 18, 18, 0, 18, 18]
[18, 18, 18, 18, 18, 18, 0, 18]
[18, 18, 18, 18, 18, 18, 18, 0]
```

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

Commit:
[c10d] Introduce a util for detecting DMA connectivity among devices

[ghstack-poisoned]

weifengpy commented on test/distributed/test_symmetric_memory.py:
```diff
@@ -90,6 +90,19 @@ def _verify_symmetric_memory(self, symm_mem):
+        self.assertEqual(connectivity.connection_type, "nvlink")
+        self.assertEqual(len(connectivity.matrix), torch.cuda.device_count())
+        for row in connectivity.matrix:
+            self.assertEqual(len(row), torch.cuda.device_count())
```
For 8 GPUs, if `self.world_size == 2`, would `len(row)` change at all?

weifengpy replied:
are we not checking value == 18 because we do not know what hardware it is ?

yifuwang replied:
> For 8 GPUs, if self.world_size == 2, would len(row) change at all?

Right. This is checking the connectivity among visible cuda devices. For checking the connectivity among ranks, we need to build another layer on top.

> are we not checking value == 18 because we do not know what hardware it is ?

Yes that's right.

weifengpy commented on torch/csrc/distributed/c10d/CudaDMAConnectivity.cpp:
```diff
@@ -0,0 +1,118 @@
+
+    std::vector<int> switch_link_count(num_devices, 0);
+    for (int i = 0; i < num_devices; ++i) {
+      for (int link = 0; link < 20; ++link) {
```
Is `20` arbitrary with no meaning or something meaningful?

yifuwang replied:
Good question. As far as I know, there's no nvml API that allows us to query the number of NVLinks. So we just increment the link id until it fails. Here we just want a reasonably large number.

I'll change this to a constant and add a comment.

weifengpy commented on torch/csrc/distributed/c10d/CudaDMAConnectivity.cpp:
```diff
@@ -0,0 +1,118 @@
+    }
+
+    return c10::make_intrusive<c10d::DMAConnectivity>(
+        c10::DeviceType::CUDA, "nvlink", std::move(matrix));
```
maybe summarize the meaning for `matrix[i][j]` ? like `matrix[i][j]>0` means blabla, or `matrix[i][j]==x` means blabla?

Commit:
Update on "[c10d] Introduce a util for detecting DMA connectivity among devices"


This PR introduces `_detect_dma_connectivity` - a utility for detecting DMA connectivity among devices.

The "DMA connectivity" in this context is more stringent than the ability to perform memory copy without CPU involvement. We define it as the ability for a device to issue load/store instructions and perform atomic operations on memory that resides on connected devices. The ability translates to the ability to run most aten GPU operations with operands backed by remote memory. `_detect_dma_connectivity` can help PyTorch and its users to determine whether certain DMA-based optimizations are possible.

`_detect_dma_connectivity` takes a `(device_type, connection_type)` pair and returns a matrix describing the connectivity. Connectivity detectors are statically registered on a `(device_type, connection_type)` basis. This PR implements the detector for `(CUDA, "nvlink")`. Later, detectors for pairs such as `(ROCM, "infinity_fabric")` can be introduced.

Example:

```python3
>>> from torch._C._autograd import DeviceType
>>> from torch._C._distributed_c10d import _detect_dma_connectivity
>>> connectivity = _detect_dma_connectivity(DeviceType.CUDA, "nvlink")
>>> for row in connectivity.matrix:
...     print(row)
...
[0, 18, 18, 18, 18, 18, 18, 18]
[18, 0, 18, 18, 18, 18, 18, 18]
[18, 18, 0, 18, 18, 18, 18, 18]
[18, 18, 18, 0, 18, 18, 18, 18]
[18, 18, 18, 18, 0, 18, 18, 18]
[18, 18, 18, 18, 18, 0, 18, 18]
[18, 18, 18, 18, 18, 18, 0, 18]
[18, 18, 18, 18, 18, 18, 18, 0]
```

cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 awgu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

yifuwang commented:
@pytorchbot merge
